docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,626 @@
+
+*Note: Image layer IDs can be obtained by running `docker history -q --no-trunc <imagename>`*
+
+##### Aufs
+
+Image layers and writable layers are stored under `/var/lib/docker/aufs/diff/<id>`.
+
+The writable layers ID is equivalent to that of the container ID.
+
+##### Devicemapper
+
+Each container and each image gets own block device. Since this driver works at the block level, it is not possible to access the layers directly without mounting them. Each container gets its own block device while running.
+
+##### Overlayfs
+
+Image layers and writable layers are stored under `/var/lib/docker/overlay/<id>`.
+
+Identical files are hardlinked between images.
+
+The image layers contain all their data under a `root` subdirectory.
+
+Everything under  `/var/lib/docker/overlay/<container-id>` are files required for running the container, including its writable layer.
+
+### Improve disk accounting
+
+Disk accounting is dependent on the storage driver in docker. A common solution that works across all storage drivers isn't available.
+
+I’m listing a few possible solutions for disk accounting below along with their limitations.
+
+We need a plugin model for disk accounting. Some storage drivers in docker will require special plugins.
+
+#### Container Images
+
+As of today, the partition that is holding docker images is flagged by cadvisor, and it uses filesystem stats to identify the overall disk usage of that partition.
+
+Isolated usage of just image layers isn’t available today.
+
+However, continuing to use the entire partition availability for garbage collection purposes in kubelet, should not affect reliability.
+We might garbage collect more often.
+As long as we do not expose features that require persisting old containers, computing image layer usage wouldn’t be necessary.
+
+In case we choose to compute the size of image layers alone, the following are some of the ways to achieve that.
+
+*Note that the strategies mentioned below are applicable in general to other kinds of storage like volumes, etc.*
+
+##### Overlayfs and Aufs
+
+1. `du`
+
+	We can list all the image layer specific directories, excluding container directories, and run `du` on each of those directories.
+
+**Pros**:
+
+* This is the least-intrusive approach.
+
+* It will work off the box without requiring any additional configuration.
+
+**Cons**:
+
+* `du` can consume a lot of cpu and memory. There have been several issues reported against the kubelet in the past that were related to `du`.
+
+* It is time consuming. Cannot be run frequently. Requires special handling to constrain resource usage - setting lower nice value or running in a sub-container.
+
+* Can block container deletion by keeping file descriptors open.
+
+
+2. Linux gid based Disk Quota
```

Comment:
nit: this `2` would be another `1` again from the [web page](https://github.com/vishh/kubernetes/blob/disk-accounting/docs/proposals/disk-accounting.md) because of the auto-rank :)